### PR TITLE
Update daphne traits so that they can be conditionally compiled with or without Send bounds

### PR DESCRIPTION
**Pull request by mendess**
The native version of this service will require that the daphne traits return `Send` futures. This however requires that the receiver of the methods be `Send`, which `DaphneWorker` is not.

To solve this problem, there are 3 solutions:
- duplicate all the traits and methods that use them as a trait bound
- `unsafe impl Send` for `DaphneWorker`
- use conditional compilation to generate `Send` and non-`Send` versions of the traits

I chose to go with the 3rd option because by conditionally compiling the traits based on a feature, the worker and the server can use the same traits but decide whether they want `Send` bounds or not.

The tradeoff of this technique is that `cargo check --all-targets` will no longer work. This is because `cargo` for every dependency finds the maximum set of features required by all dependents and then only compiles the code once. But since the enabling the feature is a breaking change for `daphne_worker` it will fail to compile when it tries to link with it, but I think this is an okay compromise for a temporary change

### CODE REVIEW

File: daphne_worker/src/roles/helper.rs [hunk 1]
```diff
@@ -23,12 +23,15 @@ use prio::codec::Encode;
 
 #[async_trait(?Send)]
 impl<'srv> DapHelper<DaphneWorkerAuth> for DaphneWorker<'srv> {
-    async fn put_helper_state_if_not_exists(
+    async fn put_helper_state_if_not_exists<Id>(
```
Comment (cjpatton) [inline daphne_worker/src/roles/helper.rs:26]: Same question as in a file above (here and below)

File: daphne/src/auth.rs [hunk 2]
```diff
@@ -51,11 +51,24 @@ impl AsRef<BearerToken> for BearerToken {
     }
 }
 
+#[cfg(feature = "send-traits")]
```
Comment (cjpatton) [inline daphne/src/auth.rs:54]: I think the `conditional_sync` module is technical enough that it needs a docucomment describing its purpose.

**Review by cjpatton — APPROVED** [review]
Looks good to me, no blocking comments.

File: daphne/src/roles/helper.rs [hunk 3]
```diff
@@ -26,394 +26,407 @@ use crate::{
...
         helper_state: &DapAggregationJobState,
-    ) -> Result<bool, DapError>;
+    ) -> Result<bool, DapError>
+    where
+        Id: Into<MetaAggregationJobId> + Send;
 
     /// Fetch the Helper's aggregation-flow state. `None` is returned if the Helper has no state
     /// associated with the given task and aggregation job.
-    async fn get_helper_state(
+    async fn get_helper_state<Id>(
```
[the remark this answers is not in the document]
Comment (mendess) [reply daphne/src/roles/helper.rs:45]: there's no guideline, when the type name is small enough I like to use a more obvious generic name

[the remark this answers is not in the document]
Comment (cjpatton) [reply daphne/src/roles/helper.rs:46]: SGTM

**Review by oliy — APPROVED** [review]

**Pull request merged**